distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Table_Helpers.enso

```diff
@@ -209,101 +206,64 @@ type Table_Column_Helper
-                ix = selector.first
-                resolved_ix = if ix < 0 then ix + col_count else ix
-
-                matches = acc.get resolved_ix []
-                acc.insert resolved_ix matches+[selector]
-
-            internal_columns.map_with_index i-> column->
-                matching_selectors = index_mapping.get i []
-                new_names = matching_selectors.map selector->
-                    selector.second
-                Pair.new column new_names
-        Column_Name_Mapping.By_Position vec ->
-            good_names = case vec.length > col_count of
-                True ->
-                    problem_builder.report_other_warning (Too_Many_Column_Names_Provided.Error (vec.drop (First col_count)))
-                    vec.take (First col_count)
-                False -> vec
-
-            new_names = good_names.map (name-> [name]) . pad col_count []
-            internal_columns.zip new_names Pair.new
+       the original table and a new name proposed for it.
+
+       The list can be empty if the column remains unchanged.
+       The list may also contain duplicate input columns.
+    columns_with_new_names = case mapping of
+        vec : Vector ->
+            col_count = internal_columns.length
+            good_names = if vec.length <= col_count then vec else
+                problem_builder.report_other_warning (Too_Many_Column_Names_Provided.Error (vec.drop (First col_count)))
+                vec.take (First col_count)
+            internal_columns.take good_names.length . zip good_names
+        _ : Map ->
+            resolve_rename selector replacement = case selector of
+                name : Text -> resolve_rename (Column_Selector.By_Name name Case_Sensitivity.Sensitive False) replacement
+                ix : Integer -> if is_index_valid internal_columns.length ix then [Pair.new (internal_columns.at ix) replacement] else
+                    problem_builder.report_oob_indices [ix]
+                    []
+                Column_Selector.By_Name name case_sensitivity use_regex ->
+                    matches = match_columns name case_sensitivity use_regex internal_columns
+                    case matches.is_empty of
+                        True ->
+                            problem_builder.report_missing_input_columns [name]
+                            []
+                        False ->
+                            if use_regex.not then matches.map c-> Pair.new c replacement else
+                                pattern = Regex.compile name case_insensitive=case_sensitivity.is_case_insensitive_in_memory
+                                matches.map c-> Pair.new c (pattern.replace c.name replacement)
+                Column_Selector.Blank_Columns _ _ ->
+                    Error.throw (Illegal_Argument.Error "Cannot use `Blank_Columns` to rename.")
+            mapping.to_vector.fold [] current-> pair-> current + resolve_rename pair.first pair.second
```

Comment:
With our current Vector impl `+` is O(N) so this will become quadratic. Not a big deal since there are not that many columns usually, but maybe worth replacing with a Vector.Builder?